src/infra/core/src/ingest/fetch_service/core.rs

```diff
@@ -46,6 +47,83 @@ pub struct FetchService {
 
 ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
 
+#[derive(Error, Debug, PartialEq)]
+pub enum TemplateError {
+    #[error(transparent)]
+    NotFound(TemplateVarValueNotFoundError),
+
+    #[error(transparent)]
+    InvalidPattern(TemplateInvalidPatternError),
+}
+
+#[derive(Error, Debug, PartialEq)]
+#[error("Value for template variable '{variable}' is missing")]
+pub struct TemplateVarValueNotFoundError {
+    pub variable: String,
+}
+
+#[derive(Error, Debug, PartialEq)]
+#[error("Invalid pattern '{pattern}' encountered in string: {template}")]
+pub struct TemplateInvalidPatternError {
+    pub template: String,
+    pub pattern: String,
+}
+
+pub fn template_string<'a>(
+    s: &'a str,
+    lookup_fn: &dyn Fn(&str) -> Option<String>,
+) -> Result<String, TemplateError> {
+    let mut s = Cow::from(s);
+    let re_tpl = regex::Regex::new(r"\$\{\{([^}]*)\}\}").unwrap();
+    let re_env =
+        regex::Regex::new(r"^env\.([a-zA-Z-_0-9]+)\s?(?:\|\|\s+([a-zA-Z0-9@:%_+-.~#?&/=]+))?$")
```

Comment:
Assuming we support two types of default values:
- integral: `123`, `-123`
- single-quoted strings: `'asd!$#@#$@#'`, `''`
perhaps we can simplify the regex to `((-?[0-9]+)|('[^']*'))` i.e. instead of listing all characters that are allowed inside a string we match *everything* until we encounter a closing single quote.

Reply:
I like this more. I was not sure so decided take just a set of valid url symbols for start.